keps/sig-network/1880-service-subnet/README.md

```diff
@@ -0,0 +1,558 @@
+	// +optional
+	Free int64 `json:"free,omitempty"`
+}
+```
+
+Reserve well-known-names for the IPRanges objects, that represent current allocators service-subnet-ranges:
+
+```go
+// default allocator well-known-names
+const (
+	DefaultIPv4Allocator = "allocator.ipv4.k8s.io"
+	DefaultIPv6Allocator = "allocator.ipv6.k8s.io"
+)
+```
+
+On the controlplane controller, add a new function similar to the existing ones used for [bootraping the default kubernetes services and namespaces](https://github.com/kubernetes/kubernetes/blob/ff34bd246b26bef622d35de68f3c143055e3aac7/pkg/controlplane/controller.go#L246-L251), to create the IPRange from the vale defined in the corresponding flag with the service-iprange-subnet
+
+
+```go
+	if utilfeature.DefaultFeatureGate.Enabled(features.ClusterIPAllocatorAPI) {
+		if err := createAllocatorIfNeeded(c.AllocatorClient, c.ServiceClusterIPRange); err != nil {
+			return err
+		}
+		if utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) && c.SecondaryServiceClusterIPRange.IP != nil {
+			if err := createAllocatorIfNeeded(c.AllocatorClient, c.SecondaryServiceClusterIPRange); err != nil {
+				return err
+			}
+		}
+	}
+```
+
+This way we guarantee that the Range Objects has been created before the Services.
+
+```
+<<[UNRESOLVED IPRange API Operations]>>
+
+* IPRange objects can not be deleted
+
+* an update on the Range field can only be allowed if there are no IP addresses allocated out of the new range.
+
+* a controller will watch IPAddress operations to update the IPRange.Status.Free value with the number
+of IP addresses free in the range
+
+<<[/UNRESOLVED]>>
+```
+
+#### IP Addresses
+
+New IP Address Object that represents a ClusterIP in the cluster
+
+```go
+// IPAddress represents an IP used by Kubernetes associated to an IPRange.
+// The name of the object is the IP address decimal number, because colons
+// are not allowed and IPv6 addresses have different text representations.
+// xref: https://tools.ietf.org/html/rfc4291
+type IPAddress struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   IPAddressSpec   `json:"spec,omitempty"`
+	Status IPAddressStatus `json:"status,omitempty"`
+}
+
+// IPRangeRef contains information that points to the IPRange being used
+type IPRangeRef struct {
+	// APIGroup is the group for the resource being referenced
+	APIGroup string
+	// Kind is the type of resource being referenced
+	Kind string
+	// Name is the name of resource being referenced
+	Name string
+}
+
+// IPAddressSpec describe the attributes in an IP Address,
+type IPAddressSpec struct {
+	// Address is the text representation of the IP Address.
+	Address string `json:"address"`
+	// IPRangeRef references the IPRange associated to this IP Address.
+	IPRangeRef IPRangeRef `json:"ipRangeRef,omitempty"`
+}
+
+// IPAddressStatus defines the observed state of IPAddress.
+type IPAddressStatus struct {
+	State IPAddressState `json:"state,omitempty"`
```

Comment:
Do we really need these?  If we use ownerRefs, I think this is all self-documenting